src/softhsm.rs

```diff
@@ -0,0 +1,141 @@
+
+use anyhow::{anyhow, Result};
+use tempdir::TempDir;
+use std::process::Command;
+
+use std::{thread, time};
+
+const TEMPDIR_PREFIX: &str = "ocicrypt-rs";
+
+pub struct SoftHSMSetup {
+    pub statedir_path: String,
+    pub statedir_folder: TempDir,
+}
+
+
+// A SoftHSM setup. Construct a new instance with ::new().
+// Calling run() methods on the instance will invoke `softhsm_setup`
+// Note: Use run_softhsm_setup() first before before calling other functions.
+impl SoftHSMSetup {
+
+    // Constructs a new SoftHSMSetup instance.
+    pub fn new() -> Result<Self> {
+        // create a temporary folder (deleted when instance goes out of scope)
+        let statedir_folder = TempDir::new(TEMPDIR_PREFIX)?;
+        // string of the temporary folder's path
+        let statedir_path = statedir_folder.path().to_string_lossy().to_string();
+        Ok(SoftHSMSetup{
+            statedir_path: statedir_path,
+            statedir_folder: statedir_folder,
+        })
+    }
+
+    // Returns the path to the softhsm configuration file.
+    pub fn get_config_filename(&self) -> Result<String> {
+        Ok(format!("{}/softhsm2.conf", self.statedir_path).to_string())
+    }
+
+    // Invokes `softhsm_setup setup` and returns the public key that was
+    // displayed
+    pub fn run_softhsm_setup(&self, softhsm_setup: &String) -> Result<String> {
+        let res = Command::new(softhsm_setup)
+                          .arg("setup")
+                          .env("SOFTHSM_SETUP_CONFIGDIR", &self.statedir_path)
+                          .output()?;
+        let res = String::from_utf8(res.stdout)?;
+        match res.find("pkcs11:") {
+            Some(idx) => {
+                let trim_me: &[_] = &[' ', '\n'];
+                let res = res[idx..].trim_end_matches(trim_me);
+                return Ok(res.to_string());
+            },
+            None => return Err(anyhow!("Failed to find 'pkcs11:' in output 
+                                       from `softhsm setup`")),
```

Comment:
Need that `\` at the end of the first line otherwise the error string will have many spaces in the middle.